### ISSUE TITLE
Asynchronous Function Calls

### ISSUE THREAD

**Issue by msshroff** [fission/fission#1230]
We have a usecase where we need to run some long running code as part of the fission functions.

But we do not want client to keep hanging on the reply. We need to send HTTP-200 immediately and actually convey the output of the function to the caller asynchronously 

Similar feature is available in OpenFaaS:
https://docs.openfaas.com/reference/async/

**Fission/Kubernetes version**

<pre>
$ fission --version
$ kubectl version
</pre>

**Kubernetes platform (e.g. Google Kubernetes Engine)**

**Describe the bug**

**To Reproduce**

**Expected result**

**Actual result**

**Screenshots/Dump file**

<pre>
$ fission support dump
</pre>

**Additional context**

**Comment by life1347** [comment]
Actually, you can achieve this now. 

Let say you can write two functions A and B. Function A receives a request, which contains callback URL in request body, from the router then publish a message to the internal NATS-Streaming server. Create a mqtrigger with function B that subscribe to the topic and do some work, then reply result with callback URL.

I think what OpenFaaS does is to simplify these steps for users. But I'm fine to have this feature since it does reduce some effort for users.

@soamvasani @vishal-biyani any thoughts?

**Comment by vishal-biyani** [comment]
@life1347 Since this is a useful feature and enables a whole bunch of new use cases, I am in favor of making this a feature that is built-in Fission!

**Comment by gkainth** [comment]
Hi. 

Has this feature made much progress? I'm currently evaluating several solutions, am leaning to Fission but would like to understand timelines on this feature.

**Comment by vishal-biyani** [comment]
Hi @gkainth We don't have any specific progress on this yet but this is easily doable:

- A primary function that accepts the message and responds with an ID.
- The function places the request in a queue.
- Another function is invoked by queue - which runs for a longer duration
- After execution of the function, the response is kept in a response queue 
- The primary function now if requested for status if that job shows as finished status. Similar logic for error too.

**Comment by brnavneet** [comment]
Hi Vishal,
Can we expect this feature to be released with v1.16.0?  The alternate idea is good and will work, but we prefer this as an OOB feature.

**Comment by sanketsudake** [comment]
Just noting down link to old proposal: https://github.com/fission/fission/blob/async-requests/Documentation/wip/async-requests.md

**Comment by 62316e** [comment]
Any updates on this one?

**Comment by sanketsudake** [comment]
Shipped — closing.

First-class async invocation landed as **RFC-0024** across #3578 (phase 1), #3579 (destinations) and #3580 (DLQ CLI + KEDA scaler), merged 2026-07-14–15. Design doc: [`docs/rfc/0024-async-invocation-retries-dlq-destinations.md`](https://github.com/fission/fission/blob/main/docs/rfc/0024-async-invocation-retries-dlq-destinations.md).

This is the OpenFaaS-style behaviour the issue asked for, built in rather than hand-assembled:

```
curl -H 'X-Fission-Invoke-Mode: async' http://$ROUTER/fn-url   # -> 202 {"invocationId": "..."}
```

or set `invocationMode: async` on the HTTPTrigger so every call to that route is async.

- The router durably enqueues and returns immediately; a dispatcher delivers at-least-once with a per-function retry policy and max age.
- **Destinations** carry the result where the caller wants it — `onSuccess`/`onFailure` can chain another function or publish to a topic — which covers "convey the output to the caller asynchronously" without the caller holding the connection.
- Failures land in a **dead-letter queue** you can inspect and redrive: `fission fn dlq list/show/redrive/purge`.
- `X-Fission-Dedup-Key` collapses duplicate enqueues; KEDA can scale the dispatcher on queue depth.

Crucially, **no broker is required** — the durable queue rides the RFC-0021 statestore substrate (#3574), so the two-function + NATS + callback-URL workaround @life1347 and @vishal-biyani described in this thread is no longer needed. (If you do want pub/sub function pipelines, RFC-0027 #3583–#3585 adds `fission topic` and a zero-broker MQ provider.)

/cc @gkainth @brnavneet @62316e @gregb122 — this is the built-in feature you were waiting on.

Tracked as #3570 under epic #3566.
